internal/controller/bindings/forwarder_controller.go

```diff
@@ -112,11 +120,112 @@ func (r *ForwarderReconciler) update(ctx context.Context, epb *bindingsv1alpha1.
 		"port", epb.Spec.Port,
 	)
 
+	// Get the KubernetesOperator
+
+	operators := ngrokv1alpha1.KubernetesOperatorList{}
+	if err := r.Client.List(ctx, &operators); err != nil {
+		log.Error(err, "failed to list KubernetesOperators")
+		return err
+	}
+	if len(operators.Items) == 0 {
+		log.Info("no KubernetesOperators found")
+		return fmt.Errorf("no KubernetesOperators found")
+	}
+	op := operators.Items[0]
+
+	// Bindings should be enabled on the operator, if they aren't we can't do anything
+	if op.Spec.Binding == nil {
+		log.Info("ut oh, no binding", "op", op)
+		return fmt.Errorf("no binding")
+	}
+
+	if op.Spec.Binding.IngressEndpoint == nil {
+		log.Info("ut oh, no ingress endpoint", "op", op)
+		return fmt.Errorf("no ingress endpoint")
+	}
+
+	// Get the secret
+	secret := v1.Secret{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: op.Namespace, Name: op.Spec.Binding.TlsSecretName}, &secret); err != nil {
+		return err
+	}
+
+	keyData, hasKey := secret.Data["tls.key"]
+	certData, hasCert := secret.Data["tls.crt"]
+
+	if !hasKey || !hasCert {
+		return fmt.Errorf("missing tls.key or tls.crt")
+	}
+
+	cert, err := tls.X509KeyPair(certData, keyData)
+	if err != nil {
+		return err
+	}
+
+	tlsDialer := tls.Dialer{
+		NetDialer: &net.Dialer{
+			Timeout: 3 * time.Minute,
+		},
+		Config: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+		},
+	}
+
+	endpointURI, err := url.Parse(epb.Spec.EndpointURI)
+	if err != nil {
+		return err
+	}
+
+	host := endpointURI.Hostname()
+	port, err := strconv.Atoi(endpointURI.Port())
+	if err != nil {
+		return err
+	}
+
+	cnxnHandler := func(conn net.Conn) error {
+		defer conn.Close()
+
+		log := log.WithValues(
+			"remoteAddr", conn.RemoteAddr(),
+			"ingress", map[string]string{
+				"endpoint": *op.Spec.Binding.IngressEndpoint,
+			},
+			"binding", map[string]string{
+				"host": host,
+				"port": strconv.Itoa(port),
+			},
+		)
+
+		log.Info("Handling connnection")
+
+		ngrokConn, err := tlsDialer.Dial("tcp", *op.Spec.Binding.IngressEndpoint)
+		if err != nil {
+			log.Error(err, "failed to dial ingress endpoint")
+			return err
+		}
+
+		// Upgrade the connection to a binding connection
+		resp, err := mux.UpgradeToBindingConnection(log, ngrokConn, host, port)
+		log = log.WithValues("endpoint.id", resp.EndpointID, "proto", resp.Proto)
+		if err != nil {
+			log.Error(err, "failed to upgrade connection")
+			return err
+		}
+
+		if resp.ErrorCode != "" || resp.ErrorMessage != "" {
+			err := fmt.Errorf("%s: %s", resp.ErrorCode, resp.ErrorMessage)
+			log.Error(err, "failed to upgrade connection", "errorCode", resp.ErrorCode, "errorMessage", resp.ErrorMessage)
+			return err
```

Comment:
OOC: Does returning errors from the forwarder lead to CrashLoopBackoff or a NotReady state via its health checks?

Reply:
Not currently. The challenge I see is that if a single port forwarder is returning errors, but the rest are fine, what do we do? Since our services all use the same selectors for targeting the forwarder pods, marking them not ready if just one forwarder has an issue would lead to knocking all port forwarders offline

Reply:
I would think that having 1 Pod marked as NotReady would tell k8s Service to update the EndpointSlice which would then have our client services not talk to the Pods that are NotReady.

I think that this is what we want, no?

Reply:
We chatted in Slack about this. This is a hard problem to solve because we don't want 1 BoundEndpoint's port to take the whole Pod offline because that would affect other ports in that Pod too.

Instead we talked about ways we could check for statuses of individual Pod IPs via headless Services, but came to the conclusion that because we're relying on k8s' service network routing that we may not be able to solve for this exact case using the proxy forwarder implementation.

So we will come back to this question post launch.